File: src/bin/0002-add-two-numbers.py
# @leetup=info id=2 lang=python3 slug=add-two-numbers





class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @leetup=code


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy_head = ListNode(0)
        prev_node = dummy_head
        carry = 0
        while l1 or l2 or carry:
            digit = 0
            if l1:
                digit += l1.val
            if l2:
                digit += l2.val
            digit += carry
            carry = digit // 10
            digit = digit % 10

            new_node = ListNode(digit)
            prev_node.next = new_node
            prev_node = new_node
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

        return dummy_head.next


# @leetup=code

File: src/bin/0002-add-two-numbers.rs
// @leetup=info id=2 lang=rust slug=add-two-numbers


// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// @leetup=code

#[allow(unused_imports)]
use std::collections::*;

impl Solution {
    pub fn add_two_numbers(
        mut l1: Option<Box<ListNode>>,
        mut l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy_head = Box::new(ListNode::new(0));
        let mut head = &mut dummy_head;
        let mut carry = 0;
        while l1.is_some() || l2.is_some() || carry == 1 {
            let mut digit = 0;
            if let Some(d) = l1 {
                digit += d.val;
                l1 = d.next;
            }
            if let Some(d) = l2 {
                digit += d.val;
                l2 = d.next;
            }
            digit += carry;
            carry = digit / 10;
            digit = digit % 10;

            let node = Some(Box::new(ListNode::new(digit)));
            head.next = node;
            head = head.next.as_mut().unwrap();
        }
        dummy_head.next
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::add_two_numbers;
}
